partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt

```diff
@@ -0,0 +1,160 @@
+package org.partiql.lang.graph
+
+import org.partiql.lang.util.tail
+
+object GraphEngine {
+
+    /** Compute any [StepSpec] (an elementary node-edge-node path, in any direction),
+     *  by combining the scan methods from [Graph].
+     *  TODO: Consider making this method as *the* API of Graph,
+     *  as it is, by itself, tirned out to be sufficient, so far as a "window" in a graph
+     *  for the needs of this engine's implementation.
+     */
+    fun Graph.getMatchingSteps(spec: StepSpec): List<Triple<Graph.Node, Graph.Edge, Graph.Node>> {
+        val (dirSpec, tripleSpec) = spec
+        val result =
+            (if (dirSpec.wantUndir) this.getUndir(tripleSpec) else emptyList()) +
+                (if (dirSpec.wantLeft && dirSpec.wantRight) this.getDirectedBlunt(tripleSpec) else emptyList()) +
+                (if (!dirSpec.wantLeft && dirSpec.wantRight) this.getDirectedStraight(tripleSpec) else emptyList()) +
+                (if (dirSpec.wantLeft && !dirSpec.wantRight) this.getDirectedFlipped(tripleSpec) else emptyList())
+        return result
+    }
+
+    /** The entry point for computing a graph pattern (translated to the [MatchSpec] "plan")
+     *  against a graph. */
+    fun evaluate(graph: Graph, matchSpec: MatchSpec): MatchResult {
+        val strideResults = matchSpec.strides.map { evaluateStride(graph, it) }
+        return joinStridesOnBinders(strideResults)
+    }
+
+    private fun evaluateStride(graph: Graph, stride: StrideSpec): StrideResult {
+        if (stride.elems.size == 1) return evaluateNodeStride(graph, stride)
+
+        val plan = planStride(stride)
+        check(stride == restoreStrideSpec(plan))
+        return evaluatePlan(graph, plan)
+    }
+
+    // This is for a node-only pattern like  (g MATCH (x)). TODO: Something less ugly / more unified with the rest?
+    private fun evaluateNodeStride(graph: Graph, stride: StrideSpec): StrideResult {
+        check(stride.elems.size == 1)
+        return when (val node = stride.elems[0]) {
+            is EdgeSpec -> { check(false); /*dummy*/ StrideResult(stride, emptyList()) }
+            is NodeSpec -> StrideResult(
+                stride,
+                graph.getNodes(node.label).map { Stride(listOf(it)) }
+            )
+        }
+    }
+
+    /** Determine the order of joining the steps in a stride.
+     *  Assumes that [stride] contains a properly alternating list: n, e, n, ..., e, n.
+     *
+     *  At the moment, this just takes one straightforward order.
+     *  Conceivably, this can become something more sophisticated.
+     */
+    private fun planStride(stride: StrideSpec): StrideTree {
```

Comment:
Should this planning be lifted into GPMLTranslator? This seems to perhaps be a preliminary for of the `expansion` form section 6.3 of the GPML paper?

If the planning were done by GPML translator, then any errors encountered could be thrown prior to the start of evaulation.